packages/gatsby-source-contentful/src/source-nodes.js

```diff
@@ -319,30 +320,31 @@ export async function sourceNodes(
   const newOrUpdatedEntries = new Set()
   entryList.forEach(entries => {
     entries.forEach(entry => {
-      newOrUpdatedEntries.add(`${entry.sys.id}___${entry.sys.type}`)
+      newOrUpdatedEntries.add(generateReferenceId(entry))
     })
   })
 
   // Update existing entry nodes that weren't updated but that need reverse
   // links added.
   existingNodes
-    .filter(n => newOrUpdatedEntries.has(`${n.id}___${n.sys.type}`))
+    .filter(
+      n => n?.sys?.type && newOrUpdatedEntries.has(generateReferenceId(n))
+    )
     .forEach(n => {
-      if (foreignReferenceMap[`${n.id}___${n.sys.type}`]) {
-        foreignReferenceMap[`${n.id}___${n.sys.type}`].forEach(
-          foreignReference => {
-            // Add reverse links
-            if (n[foreignReference.name]) {
-              n[foreignReference.name].push(foreignReference.id)
-              // It might already be there so we'll uniquify after pushing.
-              n[foreignReference.name] = _.uniq(n[foreignReference.name])
-            } else {
-              // If is one foreign reference, there can always be many.
-              // Best to be safe and put it in an array to start with.
-              n[foreignReference.name] = [foreignReference.id]
-            }
+      const id = generateReferenceId(n)
+      if (foreignReferenceMap[id]) {
+        foreignReferenceMap[id].forEach(foreignReference => {
+          // Add reverse links
+          if (n[foreignReference.name]) {
+            n[foreignReference.name].push(foreignReference.id)
```

Comment:
I wonder if this breaks the unit tests // reference resolving 🤷 

https://app.circleci.com/pipelines/github/gatsbyjs/gatsby/77384/workflows/2f3c29dd-8b47-4620-ad33-5e3cea38141c/jobs/913169